### PR TITLE
[V1] Aggregate chunked prompt logprobs in model runner

### PR DESCRIPTION
**Pull request by njhill**
Addresses https://github.com/vllm-project/vllm/issues/14239

Alternative to https://github.com/vllm-project/vllm/pull/14240 avoiding tensor concatenation by pre-allocating CPU tensors and copying GPU chunks into them.

@afeldman-nm

### CODE REVIEW

File: vllm/v1/core/scheduler.py [hunk 1]
```diff
@@ -637,6 +636,8 @@ def update_from_output(
                         new_prompt_logprobs_tensors=prompt_logprobs_tensors,
                         stop_reason=request.stop_reason,
                         events=request.take_events()))
+            else:
```
Comment (afeldman-nm) [inline vllm/v1/core/scheduler.py:639]: Nit: explanatory comment, i.e. "Invariant: EngineCore returns no partial prefill outputs". This invariant was not obvious to me when I wrote the logprobs PR.

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -1201,16 +1202,50 @@ def _get_prompt_logprobs_dict(
...
+                    in_progress_dict[req_id] = logprobs_tensors
             else:
                 # This is the last chunk of prompt tokens to return.
                 num_logits = num_remaining_tokens
+                if start_idx != 0:
+                    del in_progress_dict[req_id]
                 completed_prefill_reqs.append(req_id)
+                prompt_logprobs_dict[req_id] = logprobs_tensors
+
+            if num_logits <= 0:
```
Comment (afeldman-nm) [inline vllm/v1/worker/gpu_model_runner.py:1244]: Can `num_logits` ever be `<0` under normal (non-errant) operation?

**Review by afeldman-nm — CHANGES_REQUESTED** [review]
Hi @njhill , thanks for this nice concise fix. Made a few suggestions. Remarkably it is does not seem that this PR requires unit test changes, since partial prefill was never unit-tested (woops) ;)

File: vllm/v1/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -1231,19 +1266,22 @@ def _get_prompt_logprobs_dict(
...
 
         # Remove requests that have completed prefill from the batch
         # num_prompt_logprobs_dict.
         for req_id in completed_prefill_reqs:
             del num_prompt_logprobs_dict[req_id]
 
         # Must synchronize the non-blocking GPU->CPU transfers.
-        torch.cuda.synchronize()
+        if prompt_logprobs_dict:
+            torch.cuda.synchronize()
```
[the remark this answers is not in the document]
Comment (njhill) [reply vllm/v1/worker/gpu_model_runner.py:1284]: Yes, this already isn't executed if there are no prompt logprobs.

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
Comment (njhill) [reply vllm/v1/worker/gpu_model_runner.py:1244]: Probably not, just a habit I have to write these kinds of things defensively.

Comment (afeldman-nm) [reply vllm/v1/worker/gpu_model_runner.py:1244]: Makes sense. A "gap" in my own best-practices is that I never know whether to code defensively, vs when to let things fail in unanticipated scenarios so that we know those scenarios are happening.

File: vllm/v1/worker/gpu_model_runner.py [hunk 4]
```diff
@@ -1201,16 +1202,50 @@ def _get_prompt_logprobs_dict(
...
+                if start_idx != 0:
+                    del in_progress_dict[req_id]
                 completed_prefill_reqs.append(req_id)
+                prompt_logprobs_dict[req_id] = logprobs_tensors
+
+            if num_logits <= 0:
+                # This can happen for the final chunk if we prefilled exactly
+                # (num_prompt_tokens - 1) tokens for this request in the prior
+                # step. There are no more prompt logprobs to produce.
+                continue
```
[the remark this answers is not in the document]
Comment (afeldman-nm) [reply vllm/v1/worker/gpu_model_runner.py:1248]: See my DM

**Review by afeldman-nm — APPROVED** [review]

File: vllm/v1/worker/gpu_input_batch.py [hunk 5]
```diff
@@ -197,6 +198,9 @@ def __init__(
         # that are currently in the prefill phase.
         self.num_prompt_logprobs: dict[str, int] = {}
 
+        # To accumulate prompt logprobs tensor chunks across prefill steps.
+        self.in_progress_prompt_logprobs_cpu: dict[str, LogprobsTensors] = {}
```
Comment (robertgshaw2-redhat) [inline vllm/v1/worker/gpu_input_batch.py:202]: This need to be added to `remove_request`

Right now we have a memory leak if a request is aborted during a parital prefill since the `del self.in_progress_prompt_logprobs_cpu` is only called once we generate the first token for a partial prefill

File: vllm/v1/worker/gpu_model_runner.py [hunk 6]
```diff
@@ -1205,16 +1206,51 @@ def _get_prompt_logprobs_dict(
...
+            start_tok = start_idx + 1
             num_remaining_tokens = num_prompt_tokens - start_tok
-            if num_tokens < num_remaining_tokens:
+            if num_tokens <= num_remaining_tokens:
                 # This is a chunk, more tokens remain.
+                # In the == case, there are no more prompt logprobs to produce
+                # but we want to defer returning them to the next step where we
+                # have new generated tokens to return.
                 num_logits = num_tokens
+                if start_idx == 0:
```
Comment (robertgshaw2-redhat) [inline vllm/v1/worker/gpu_model_runner.py:1238]: We can avoid special casing `if start_idx == 0` and `start_idx != 0` if we:

- move `in_progress_dict[req_id] = logprobs_tensors` under the `if not logprobs_tensors` branch above
- move `del in_progress_dict[req_id]` under the `for req_id in completed_prefill_reqs:` loop below

This will just make the code easier to follow at the expense of slightly more work. The design goal of this is "simple" over "optimal"

**Comment by robertgshaw2-redhat** [comment]
There are two comments:
* There is a memory leak if requests are aborted during partial prefills with prompt logprobs
* I think we can simplify some special casing (this is nit)

File: vllm/v1/worker/gpu_input_batch.py (hunk 5 above)
Comment (njhill) [reply vllm/v1/worker/gpu_input_batch.py:202]: good catch, thanks

File: vllm/v1/worker/gpu_model_runner.py (hunk 6 above)
Comment (njhill) [reply vllm/v1/worker/gpu_model_runner.py:1238]: sure, I guess I was just avoiding adding/removing in the common non-chunked case but you're probably right that simpler code is a better option

Comment (robertgshaw2-redhat) [reply vllm/v1/worker/gpu_model_runner.py:1238]: yeah, its a small nit but i think making this as simple as possible is a good idea

**Review by robertgshaw2-redhat — APPROVED** [review]

**Comment by robertgshaw2-redhat** [comment]
LGTM thanks!
